src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.SectionNameMatching.cs

```diff
@@ -108,10 +112,17 @@ public bool IsMatch(string s)
                 numberRangePairs.Free();
                 return null;
             }
+
             sb.Append('$');
-            return new SectionNameMatcher(
-                new Regex(sb.ToString(), RegexOptions.Compiled),
-                numberRangePairs.ToImmutableAndFree());
+
+            var pattern = sb.ToString();
+            if (!RegexMap.TryGetValue(pattern, out var regex))
+            {
+                regex = new Regex(pattern, RegexOptions.Compiled);
+                regex = RegexMap.GetOrAdd(pattern, regex);
```

Comment:
Rather than this being regex -> Regex, can we make the cache be pattern -> Regex? Then we can check this at the very very top and avoid doing all the section name to regex conversion.